### PR TITLE
Use Wikidata IRIs for processing fields (#1)

### PR DESCRIPTION
**Pull request by ross-spencer**
Corrects the issue described in #153 so that language choices do not impact the processing of Wikidata signature definitions into a signature file. 

Notes:

* Contains a handful of smaller code-style changes in separate commits so might suggest rebase before merging.
* Wikidata definitions file becomes 2.0.0 because of the replacement of some fields, e.g. `encodingLabel` replaced with `encoding`.
* Will need to update some of the details on https://github.com/richardlehane/siegfried/wiki/Wikidata-identifier#harvesting 

Connected to https://github.com/richardlehane/siegfried/issues/153

### CODE REVIEW

File: cmd/roy/data/wikidata/wikidata-test-definitions [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "head" : {
-    "vars" : [ "uri", "uriLabel", "puid", "extension", "mimetype", "encodingLabel", "referenceLabel", "date", "relativityLabel", "offset", "sig" ]
+    "vars" : [ "uri", "uriLabel", "puid", "extension", "mimetype", "encoding", "referenceLabel", "date", "relativity", "offset", "sig" ]
```
Comment (ross-spencer) [inline cmd/roy/data/wikidata/wikidata-test-definitions:3]: Removing label from encoding and relativity fields will give us the IRIs for those values instead. The tests are updated here to verify this change. A few new entries have been added to demonstrate Siegfried results in other encodngs. 

Mappings to pay attention to below: 

* IRI ending Q82828 - Hexadecimal.
* IRI ending Q35436009 - beginning of file.
* IRI ending Q1148480 - end of file.

File: cmd/sf/wikidata_test.go [hunk 2]
```diff
@@ -61,6 +61,7 @@ func setupWikidata(pronomx bool) error {
 // identificationTests provides our structure for table driven tests.
 type identificationTests struct {
 	fname          string
+	label          string
```
Comment (ross-spencer) [inline cmd/sf/wikidata_test.go:64]: Label added to testing structures to assert the output in our different encodings works as anticipated. Struct data updated below.

File: pkg/config/wikidata.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ var wikidata = struct {
 	gzip:         "Q27824060",
 	tar:          "Q283579",
 	warc:         "Q10287816",
-	definitions:  "wikidata-definitions-1.0.0",
+	definitions:  "wikidata-definitions-2.0.0",
```
Comment (ross-spencer) [inline pkg/config/wikidata.go:70]: Definitions version updated here. TODO: come back to the implications...

**Comment by ross-spencer** [comment]
NB. Can delete this comment once merged, but this is the text for the Wiki:

```
The Wikidata identifier harnesses the file format signatures in Wikidata that can be made to be compatible with Siegfried. Data can be downloaded to create a new identifier and that identifier used to scan the objects in your collection. 

For **developers** the exposed aspects of the integration API have been documented and can be viewed on [go.dev](https://pkg.go.dev/github.com/richardlehane/siegfried/pkg/wikidata). As well as the rest of the [Siegfried](https://pkg.go.dev/github.com/richardlehane/siegfried/) interfaces. 

For **users**, then a basic understanding of the Roy tool and [how to build identifiers](https://github.com/richardlehane/siegfried/wiki/Building-a-signature-file-with-ROY) will be helpful. Understanding Siegfried's [identification capabilities](https://github.com/richardlehane/siegfried/wiki/Identifying-file-formats) will also be useful. 

## Overview

Given prior knowledge of Siegfried and how to configure its defaults, the commands below can be used to make use of the Wikidata integration. 

### Harvesting

Harvest a Wikidata signature file as follows: `roy harvest -wikidata`

The file which this creates can be found at `$HOME/<user>/siegfried/wikidata/wikidata-definitions-<wikidata-version>`.

The SPARQL query used to generate version 1.0.0 of the Wikidata identifier should be in the Wikidata SPARQL module [here](https://github.com/richardlehane/siegfried/blob/master/pkg/config/internal/wikidatasparql/sparql.go).

Version 2.0.0. of the identifier uses the following query:

```sparql
# Return all file format records from Wikidata.
#
select distinct ?uri ?uriLabel ?puid ?extension ?mimetype ?encoding ?referenceLabel ?date ?relativity ?offset ?sig
where
{
  ?uri wdt:P31/wdt:P279* wd:Q235557.               # Return records of type File Format.
  optional { ?uri wdt:P2748 ?puid.      }          # PUID is used to map to PRONOM signatures proper.
  optional { ?uri wdt:P1195 ?extension. }
  optional { ?uri wdt:P1163 ?mimetype.  }
  optional { ?uri p:P4152 ?object;                 # Format identification pattern statement.
    optional { ?object pq:P3294 ?encoding.   }     # We don't always have an encoding.
    optional { ?object ps:P4152 ?sig.        }     # We always have a signature.
    optional { ?object pq:P2210 ?relativity. }     # Relativity to beginning or end of file.
    optional { ?object pq:P4153 ?offset.     }     # Offset relative to the relativity.
    optional { ?object prov:wasDerivedFrom ?provenance;
       optional { ?provenance pr:P248 ?reference;
                              pr:P813 ?date.
                }
    }
  }
  service wikibase:label { bd:serviceParam wikibase:language "[AUTO_LANGUAGE], <<lang>>, en". }
}
order by ?uri
```

#### Changing the WikiBase URI

If you have access to another [WikiBase](https://wikiba.se/) implementation that can serve Wikidata compatible file format information you can change the URI from where the information is harvested as follows:

* `roy harvest -wikidataendpoint http://sparql.example.com`

#### Changing the Wikidata results language

The Wikidata results can be returned in a different language. Where a translation is available for a file format or a format's native language was something other than English this can be useful for finding the information you need. A different language can be returned using the following command:

* `/roy harvest -wikidata -lang <two-letter-language-code>`

E.g. German (DE)

* `/roy harvest -wikidata -lang de`

An example format with a German translation is "Microsoft Shortcut" or "Dateiverknüpfung": http://www.wikidata.org/entity/Q1109779

E.g. 

```yaml
filename : 'shortcut.lnk'
filesize : 8
modified : 2021-04-18T17:07:20+02:00
errors   : 
matches  :
  - ns      : 'wikidata'
    id      : 'Q1109779'
    format  : 'Dateiverknüpfung'
    URI     : 'http://www.wikidata.org/entity/Q1109779'
    mime    : 'application/x-ms-shortcut'
    basis   : 'byte match at 0, 8'
    source  : 'Wikidata reference is empty'
    warning : 'extension mismatch'
```

Where a translation doesn't exist for a file format, the translation will fall back, for now, to English (EN). 

### Building

Siegfried's binary representation of a signature file is called an Identifier. The identifier must be compiled. It can be compiled using many different combinations discussed in the Roy documentation. We will focus on Wikidata with PRONOM and Wikidata without PRONOM. 

#### PRONOM

Wikidata will be build with PRONOM by default. What this does is look for PRONOM identifiers in the Wikidata dataset. Those identifiers might not have a signature associated with them where PRONOM does. We supercharge Wikidata by making use of a set of both information sources. 

* `roy build -wikidata`

#### No PRONOM

To build a Wikidata identifier without PRONOM, for example, to test your Wikidata developed signatures more easily:

* `roy build -wikidata -nopronom`

#### Logging

Roy will output information specific to the Wikidata identifier which can be helpful to you to see the amount of information in Wikidata that you can expect to use. An example below shows that there are 192 records with signatures. This means the Wikidata identifier on its own can identify up to 192 file formats through binary pattern matching. 

```json
{
  "AllSparqlResults": 13187,
  "CondensedSparqlResults": 4582,
  "SparqlRowsWithSigs": 2927,
  "RecordsWithPotentialSignatures": 196,
  "FormatsWithBadHeuristics": 4,
  "RecordsWithSignatures": 192,
  "MultipleSequences": 11,
  "AllLintingMessages": [
    "Use the `-wikidataDebug` flag to build the identifier to see linting messages"
  ],
  "AllLintingMessageCount": 134,
  "RecordCountWithLintingMessages": 116
}
```
Linting can help you in developing file format signatures in Wikidata and identifying errors in that process. It is described below. 

#### Linting

It can be helpful to Wikidata signature developers to identify potential issues in Wikidata signatures. The technique used in version `1.9.x` of Siegfried is not a perfect technique. We anticipate greater schema checking against the Wikidata data source using [ShEx](https://shex.io/) in time.

For now, when you build using the following parameters, you will see additional "linting" information to help you identify records in Wikidata that can be improved with your attention:

* `roy build -wikidatadebug`

```json
  "AllLintingMessages": [
    "Linting: WARNING no encoding: URI: http://www.wikidata.org/entity/Q4839791 Critical: false",
    "Linting: WARNING no provenance: URI: http://www.wikidata.org/entity/Q4839791 Critical: false",
    "Linting: WARNING no provenance date: URI: http://www.wikidata.org/entity/Q98843338 Critical: false",
    "Linting: ERROR bad heuristic: URI: http://www.wikidata.org/entity/Q1109779 Critical: true",
    "Linting: ERROR blank node returned for offset: URI: http://www.wikidata.org/entity/Q26546575 Critical: false",
    "Linting: WARNING no relativity: URI: http://www.wikidata.org/entity/Q939636 Critical: false",
  ],
```
We can go into more information about linting issues:

##### ERROR bad heuristic

A bad heuristic means that information vital to understanding a signature is missing. Wikidata is the first place to look for this inconsistency. An example of a bad heuristic might be a file format listed with two BOF sequences but no offset to describe how one is related to the other. The weakness might be in the code where the code does not demonstrate enough complexity to work with the data that is available to it. If you believe a heuristic can be created for the information in Wikidata please open a new Siegfried issue. 

##### ERROR blank node returned

A blank node can be returned for any field that Roy/Siegfried anticipates using. A blank node error is returned for a field that has been deliberately listed in Wikidata but for which there has been no value supplied, e.g. the author recognizes there should be something but does not know what that something is. Roy cannot work with this field/value as it is incomplete. The best way to remedy this is to complete the record in Wikidata.

##### WARNING no encoding

A no-encoding error exists in Roy at present because Wikidata can encode information using multiple signature encoding, e.g. `hexadecimal`, `ascii`. If an encoding isn't specified Roy will try to parse or convert the data to `hexadecimal`. If it works no error will be thrown and we can use the signature. This error is a signal to the signature developer to rectify the issue in the Wikidata record. 

##### WARNING no provenance

A no provenance error indicates that the signature information in Wikidata has no listed reference. A default value will be used in the output by Siegfried. The remedy is to attempt to find a suitable provenance for the information in Wikidata and edit the record directly. 

##### WARNING no provenance date

A no provenance date error indicates that the signature information in Wikidata has no listed date for its reference. No value will be used by Siegfried. The remedy is to attempt to find a suitable provenance date for the information in Wikidata and edit the record directly. 

##### WARNING no relativity

A no relativity error indicates that the signature information in Wikidata has no listed relativity value, e.g. it is not listed as `BOF` (beginning of file) or `EOF` (end of file). A default of `BOF` value will be used in the output by Siegfried. The remedy is to attempt to find a suitable provenance for the information in Wikidata and edit the record directly. 

### Inspect

Signatures can be inspected on a case-by-case basis. To lookup a Wikidata identifier to see the signature compiled into the identifier you can do the following

* `roy inspect -wikidata <Wikidata-QID>`

E.g. [FLAC](https://www.wikidata.org/wiki/Q219848) (Free Lossless Audio Codec)

* `roy inspect -wikidata Q27881556`

```yaml
FORMAT INFO: NAME: 'FLAC'
MIMETYPE: 'AUDIO/X-OGG; AUDIO/X-FLAC; AUDIO/FLAC'
SOURCES: 'GARY KESSLER'S FILE SIGNATURE TABLE (SOURCE DATE: 2017-08-08) PRONOM (OFFICIAL (FMT/279))'
QID: (Q27881556)
globs: *.flac, *.oga
sigs: (B:0 seq "fLaC\x00\x00\x00\"")
      (B:0..4 seq "fLaC\x00\x00\x00\"")
superiors: none
```

### Scanning

One an identifier is built, Siegfried does not require any special invocation to use Wikidata. A standard command might be `sf <your-file-name>`. The result, e.g. for `img.bmp` will look something like as follows:

```yaml
---
siegfried   : 1.9.1
scandate    : 2020-11-15T22:24:56-05:00
signature   : default.sig
created     : 2020-11-15T22:24:35-05:00
identifiers : 
  - name    : 'wikidata'
    details : 'wikidata-definitions-1.0.0 (2020-11-15)'
---
filename : 'img.bmp'
filesize : 35
modified : 2020-11-15T22:26:09-05:00
errors   : 
matches  :
  - ns       : 'wikidata'
    id       : 'Q27596325'
    format   : 'Windows Bitmap, version 4'
    URI      : 'http://www.wikidata.org/entity/Q27596325'
    mime     : 
    basis    : 'extension match bmp; byte match at 0, 35'
    source   : 'PRONOM (Wikidata) (source date: 2017-08-08)'
    warning  : 
    software : 
```

## Terminology

* QID: [Identifiers in Wikidata](https://www.wikidata.org/wiki/Q43649390) have a `Q` prefix. `Q`-Identifier.
* Wikibase: A semantic web platform on which Wikidata is built.
* Roy: Tool to harvest and build (compile) Siegfried identifiers from sources such as Wikidata.
* Siegfried: A consumer of the Roy built identifier, returning file-format information based on binary pattern matching. 
```

File: cmd/roy/data/wikidata/wikidata-test-definitions (hunk 1 above)
Comment (ross-spencer) [reply cmd/roy/data/wikidata/wikidata-test-definitions:3]: NB. I have added annotations to this file to help with any understanding in future, but I am not wedded to those annotations moving forward.

**Review by richardlehane — APPROVED** [review]
sorry to take so long to get to this Ross, but it looks great

**Pull request merged**